File: Cpp/lista_2/ArabRzym.cpp
#include "ArabRzym.h"
#include <cctype>

int liczbyArab[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
std::string liczbyRzym[] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

ArabRzymException::ArabRzymException(std::string msg) {
    message = msg;
}

bool ArabRzym::isValidRzym(std::string input) {
    for (char rzym : input) {
        rzym = toupper(rzym);
        if (rzym != 'I' && rzym != 'V' && rzym != 'X' && rzym != 'L' &&
            rzym != 'C' && rzym != 'D' && rzym != 'M') {
            return false;
        }
    }
    return true;
}

bool ArabRzym::isValidArab(std::string input) {
    for (char arab : input) {
        if (!isdigit(arab)) return false;
    }
    return true;
}

int ArabRzym::getArabic(std::string rzymDigit) {
    int arabValue = 0;
    for (int j = 0; j < 13; j++) {
        if (liczbyRzym[j] == rzymDigit) {
            arabValue = liczbyArab[j];
        }
    }
    return arabValue;
}

std::string ArabRzym::getRzym(int arabicDigit) {
    std::string rzymValue = "";
    for (int j = 0; j < 13; j++) {
        if (liczbyArab[j] == arabicDigit) {
            rzymValue = liczbyRzym[j];
        }
    }
    return rzymValue;
}

std::string ArabRzym::arab2rzym(int arab) {
    if (arab <= 0 || arab > 4000) {
        throw ArabRzymException("Liczba z poza zakresu.");
    }

    std::string result = "";
    while (arab != 0) {
        for (int i = 0; i < 13; i++) {
            while (liczbyArab[i] <= arab) {
                arab -= liczbyArab[i];
                result += liczbyRzym[i];
            }
        }
    }

    return result;
}


int ArabRzym::rzym2arab(std::string rzym) {
    if (rzym.empty()) {
        throw ArabRzymException("Wartość pusta.");
    }

    for (int i = 0; i < rzym.size(); i++) {
        rzym[i] = toupper(rzym[i]);
    }

    if (!isValidRzym(rzym)) {
        throw ArabRzymException("Nieprawidłowa liczba rzymska: " + rzym);
    }

    int result = 0;
    for (int i = 0; i < rzym.length(); i++) {
        std::string rzym_elem(1, rzym[i]);
        int arabValue = getArabic(rzym_elem);

        if (i != rzym.length() - 1) {
            std::string possible_rzym = rzym_elem + rzym[i + 1];
            if (getArabic(possible_rzym) != 0) {
                arabValue = getArabic(possible_rzym);
                i++;
            }
        }
        result += arabValue;
    }

    if (arab2rzym(result) != rzym) {
        throw ArabRzymException("Nieprawidłowa liczba rzymska: " + rzym);
    }

    return result;
}

File: Java/lista_2/src/ArabRzym.java
import java.util.*;

class ArabRzymException extends Exception {
    public ArabRzymException(String message) {
        super(message);
    }
}

public class ArabRzym {
    public static boolean isValidRoman(String input) {
        return input != null && input.matches("^[IVXLCDM]+$");
    }

    public static boolean isValidArabic(String input) {
        return input != null && input.matches("^\\d+$");
    }

    private static final int[] liczbyArab = {
            1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1
    };

    private static final String[] liczbyRzym = {
            "M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"
    };

    private static int getArabic(String rzymDigit) {
        int arabValue = 0;
        for (int j = 0; j < liczbyRzym.length; j++) {
            if (liczbyRzym[j].equals(rzymDigit)) {arabValue = liczbyArab[j];}
        }
        return arabValue;
    }

    private static String getRzym(int arabicDigit) {
        String rzymValue = "";
        for (int j = 0; j < liczbyRzym.length; j++) {
            if (liczbyArab[j] == arabicDigit) {rzymValue = liczbyRzym[j];}
        }
        return rzymValue;
    }

    public static int rzym2arab(String rzym) throws ArabRzymException {
        if (rzym == null || rzym.isEmpty()) {
            throw new ArabRzymException("Wartość nie może być pusta.");
        }

        rzym = rzym.toUpperCase();

        if (!isValidRoman(rzym)) {
            throw new ArabRzymException("Nieprawidłowa liczba rzymska: " + rzym);
        }

        int wynik = 0;
        for (int i = 0; i < rzym.length(); i++) {
            String rzym_elem = String.valueOf(rzym.charAt(i));
            int arabValue = getArabic(rzym_elem);

            if (i != rzym.length() - 1) {
                String possible_rzym = rzym_elem + rzym.charAt(i + 1);
                if (getArabic(possible_rzym) != 0) {
                    arabValue = getArabic(possible_rzym);
                    i += 1;
                }
            }
            wynik += arabValue;
        }
        return wynik;
    }

     public static String arab2rzym(int arab) throws ArabRzymException {
         String result = "";

         while (arab != 0){
             for (int i = 0; i < liczbyArab.length; i++) {
                 while (liczbyArab[i] <= arab){
                     arab -= liczbyArab[i];
                     result += getRzym(liczbyArab[i]);
                 }
             }
         }

         return result;
     }
}